src/main/java/uk/gov/ch/oauth/nonce/NonceGenerator.java

```diff
@@ -0,0 +1,48 @@
+package uk.gov.ch.oauth.nonce;
+
+import java.security.SecureRandom;
+import java.util.Random;
+import org.apache.commons.codec.binary.Base64;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.ch.oauth.session.SessionFactory;
+import uk.gov.companieshouse.session.Session;
+import uk.gov.companieshouse.session.SessionKeys;
+
+@Component
+public class NonceGenerator implements INonceGenerator {
+
+    private final Random random = new SecureRandom();
+
+    @Autowired
+    SessionFactory sessionFactory;
+
+    /**
+     * Generates a nonce, adds it to the session and then returns the nonce.
+     *
+     * @param session takes a session and sets the unique key in the data. If the session is null,
+     * generates a new one.
+     * @return Base64 encoded unique key
+     */
+    @Override
+    public String setNonceForSession(Session session) {
```

Comment:
Might be better names `addNonceToSession`

Reply:
Fine.